### PR TITLE
[SPARK-47764][FOLLOW-UP] Change to use ShuffleDriverComponents.removeShuffle to remove shuffle properly

### PR DESCRIPTION
**Pull request by bozhang2820**
### What changes were proposed in this pull request?
This is a follow-up for https://github.com/apache/spark/pull/45930, where we introduced ShuffleCleanupMode and implemented cleaning up of shuffle dependencies. 

There was a bug where `ShuffleManager.unregisterShuffle` was used on Driver, and in non-local mode it is not effective at all. This change fixed the bug by changing to use `ShuffleDriverComponents.removeShuffle` instead.

### Why are the changes needed?
This is to address the comments in https://github.com/apache/spark/pull/45930#discussion_r1584223064

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Updated unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Could you re-trigger CI, @bozhang2820 ?

**Comment by dongjoon-hyun** [comment]
I converted this to `Draft` PR because the CI is broken and the PR title has `[WIP]`.

**Comment by bozhang2820** [comment]
> Could you re-trigger CI, @bozhang2820 ?

Sorry for the late reply @dongjoon-hyun.. Updated the test and removed the WIP tag. 

It seems that the current test failure is unrelated to this change?

**Comment by abellina** [comment]
@bozhang2820 this PR is needed for the feature to work in my tests, I am just wondering what the status of this is.

**Review by mridulm — COMMENTED** [review]
Looks reasonable - can you add a test for this ?
We can use `local` vs `local-cluster` to test the expected behavior

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 1]
```diff
@@ -176,7 +176,13 @@ object SQLExecution extends Logging {
             shuffleIds.foreach { shuffleId =>
               queryExecution.shuffleCleanupMode match {
                 case RemoveShuffleFiles =>
-                  SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  // Do not unregister the shuffle on MapOutputTracker here to trigger stage retry.
+                  // Otherwise, downstream tasks will fail with MetadataFetchFailedException.
+                  if (sc.isLocal) {
+                    SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  } else {
+                    sc.shuffleDriverComponents.removeShuffle(shuffleId, false)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:184]: does this work for local mode?

Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:184]: Yes `ShuffleDriverComponents.removeShuffle` should work for both local an non-local mode. Will update the code here.

**Comment by bozhang2820** [comment]
> Looks reasonable - can you add a test for this ? We can use `local` vs `local-cluster` to test the expected behavior

I tried adding a unit test for local-cluster mode but found it a bit difficult:
1. The cleanup is done in a best-effort basis, and in local-cluster mode the shuffle blocks are less likely to be cleaned up in time. 
2. There seems to be no API to get the MigratableResolver on executors. 

What do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 2]
```diff
@@ -177,7 +177,9 @@ object SQLExecution extends Logging {
             shuffleIds.foreach { shuffleId =>
               queryExecution.shuffleCleanupMode match {
                 case RemoveShuffleFiles =>
-                  SparkEnv.get.shuffleManager.unregisterShuffle(shuffleId)
+                  // Do not unregister the shuffle on MapOutputTracker here to trigger stage retry.
+                  // Otherwise, downstream tasks will fail with MetadataFetchFailedException.
+                  sc.shuffleDriverComponents.removeShuffle(shuffleId, Utils.isTesting)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:182]: This cleanup is not critical, I think the second parameter `blocking` can always be false

Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:182]: The unit tests would be flaky if we use false here.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:182]: oh I see, can we add some comments to explain it?

Comment (bozhang2820) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:183]: Sure. Done.

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @bozhang2820, @cloud-fan , @mridulm .
Merged to master for Apache Spark 4.0.0-preview2.

**Comment by dongjoon-hyun** [comment]
For the test cases, please file a JIRA issue not to forget because it seems to need more time, @bozhang2820 .

**Pull request closed without merging**